### PR TITLE
Action List: Sorting

### PR DESCRIPTION
**Pull request by ElliotTheHuman**
### :sparkles: Changes

- Updated Action List components to support sorting
- Added utility function that generates sortable Action List off of an array of data objects and an array of column-metadata objects

### :white_check_mark: Requirements

- [x] Includes test coverage for all changes
- [x] Build and tests are passing

### :camera: Screenshots
sorting

### CODE REVIEW

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx [hunk 1]
```diff
@@ -24,14 +24,56 @@
 
  */
 
+import React, { FC, useContext } from 'react'
 import styled, { css } from 'styled-components'
-import { flexbox } from '@looker/design-tokens'
+import { ActionListContext } from '../ActionListContext'
+import { Icon } from '../..'
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Resolved in https://github.com/looker-open-source/components/pull/668/commits/90fbb85ffc4800426f225ba1cc62303e8426b26e

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx [hunk 2]
```diff
@@ -24,14 +24,56 @@
...
+        doSort(id, 'desc')
+      }
+    }
+  }
+
+  return (
+    <div
+      className={className}
+      onClick={handleClick}
+      style={{ alignItems: 'center', display: 'flex' }}
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Resolved in https://github.com/looker-open-source/components/pull/668/commits/90fbb85ffc4800426f225ba1cc62303e8426b26e.

However, this required passing columns' canSort property as a prop (so that we could conditionally set "cursor" to "pointer" based on the prop received).

[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Wondering whether it makes sense or not to remove `ActionListContext` for now since it provides only two things to the `ActionListHeaderColumn`:
1. A column's current "sortDirection"
2. The "doSort" callback

I realize now that it's fairly easy to pass all column metadata (including the data above) directly as props from the `ActionList` to the `ActionListHeaderColumn`. Which begs the question whether we should also pass the above in as props as well.

[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: One idea might be to just have the `ActionListHeaderColumn` accept an object of type `ActionListColumn`.

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx [hunk 3]
```diff
@@ -24,14 +24,56 @@
...
+    >
+      {children}
+      {columnInfo && columnInfo.sortDirection ? (
+        <Icon
+          ml={columnInfo.type === 'string' ? 'xxsmall' : ''}
+          mr={columnInfo.type === 'number' ? 'xxsmall' : ''}
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Hitting this error when attempting to use the `&&` operator:
```
Type 'false' is not assignable to type 'string | number | (string | number | null)[] | { [key: string]: ReactText; } | undefi...
```
Is there a way to get around this error while still using `ml` and `mr`?

[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Hitting an error when attempting to use `&&` operator:
```
Type 'false' is not assignable to type 'string | number | (string...
```

Looks like something like this tricks typescript into accepting `false` (since it gets stringified), but wondering if it's a little too wordy:
```
ml={`${columnInfo.type === 'string' && 'xxsmall'}`}
```

File: packages/components/src/ActionList/utils/createSortableActionList.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+import {
+  ActionList,
+  ActionListColumns,
+  ActionListItem,
+  ActionListItemColumn,
+  ActionListItemAction,
+} from '..'
+
+export const createSortableActionList = (
+  data: any[],
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/createSortableActionList.tsx]: Resolved (I think) in https://github.com/looker-open-source/components/pull/668/commits/492314c9233b34b97d68f4af78c830aa80d73e32.

Added an `ActionListData` type which is equal to "Record<string, string | number>".

File: packages/components/src/ActionList/utils/createSortableActionList.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+      if (upperCasedStringA < upperCasedStringB) return -1
+      if (upperCasedStringA > upperCasedStringB) return 1
+      return 0
+    }
+
+    if (targetColumn && targetColumn.type === 'number') {
+      if (sortDirection === 'desc') {
+        sortedData.sort((a, b) => b[id] - a[id])
+        targetColumn.sortDirection = 'desc'
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/createSortableActionList.tsx]: Resolved in https://github.com/looker-open-source/components/pull/668/commits/65d55bcf75001726bfd5b7360230f3c8e6bc5d8a

File: packages/components/src/ActionList/ActionListItem.tsx [hunk 6]
```diff
@@ -91,6 +91,7 @@ export const ActionListItem = styled(ActionListItemInternal)`
   &:focus,
   &:hover {
     box-shadow: ${({ theme, onClick }) => onClick && theme.shadows[2]};
+    cursor: ${({ onClick }) => (onClick ? 'pointer' : null)};
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListItem.tsx]: Resolved in https://github.com/looker-open-source/components/pull/668/commits/65d55bcf75001726bfd5b7360230f3c8e6bc5d8a

File: packages/components/src/ActionList/utils/createSortableActionList.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    columns.forEach(column => delete column.sortDirection)
+
+    const stringComparator = (stringA: string, stringB: string) => {
+      const upperCasedStringA = stringA.toUpperCase()
+      const upperCasedStringB = stringB.toUpperCase()
+
+      if (upperCasedStringA < upperCasedStringB) return -1
+      if (upperCasedStringA > upperCasedStringB) return 1
+      return 0
+    }
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/createSortableActionList.tsx]: Resolved in https://github.com/looker-open-source/components/pull/668/commits/65d55bcf75001726bfd5b7360230f3c8e6bc5d8a

File: packages/components/src/ActionList/utils/createSortableActionList.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  ActionList,
+  ActionListColumns,
+  ActionListItem,
+  ActionListItemColumn,
+  ActionListItemAction,
+} from '..'
+
+export const createSortableActionList = (
+  data: any[],
+  setData: (newData: any[]) => void,
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/createSortableActionList.tsx]: I think having this function handle the state management of the created elements makes the function parameter cleaner! Based on what I know, that means this function will need to be converted into a hook.

Discussed over Slack, but the most elegant plan seems to be refactoring this function into a hook in a separate PR (which then allows for `useState` calls). In which case, I _believe_ this comment is safe to resolve for now.

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukelooker) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Arg... my bad, forgot that ternary will cast to `false` (which isn't valid)

```ml={columnInfo.type === 'string' ? 'xxsmall' : undefined}```

☝️  Still a minor change to use `undefined` instead of `''` since our goal is to set _nothing_ not an empty string :)

File: packages/components/src/ActionList/utils/useActionListSortManager.tsx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+} from '..'
+
+const stringComparator = (stringA: string, stringB: string) => {
+  const upperCasedStringA = stringA.toUpperCase()
+  const upperCasedStringB = stringB.toUpperCase()
+
+  if (upperCasedStringA < upperCasedStringB) return -1
+  if (upperCasedStringA > upperCasedStringB) return 1
+  return 0
+}
```
Comment (lukelooker) [inline packages/components/src/ActionList/utils/useActionListSortManager.tsx]: Needs test coverage

Comment (lukelooker) [reply packages/components/src/ActionList/utils/useActionListSortManager.tsx]: Probably worth pulling this into it's own file so it can have a related unit test file :)

File: packages/components/src/ActionList/utils/useActionListSortManager.tsx [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+) => {
+  const [data, setData] = useState(actionListData)
+  const [columns, setColumns] = useState(actionListColumns)
+
+  const MyActions = () => (
+    <>
+      <ActionListItemAction onClick={() => alert(`You performed an action!`)}>
+        Some Action
+      </ActionListItemAction>
+    </>
```
Comment (lukelooker) [inline packages/components/src/ActionList/utils/useActionListSortManager.tsx]: Doesn't need a fragment around:

```suggestion
      <ActionListItemAction onClick={() => alert(`You performed an action!`)}>
        Some Action
      </ActionListItemAction>
```

As a follow-up to this PR we might need to think about how this would actually be useful. For someone to use this it seems this will need to be something that a developer can specify as a function – something like a prop that `useActionListSortManager` accepts:

```
actions: (item: ActionListData) => ReactNode
```

And then is implemented as:

```
const actionGenerator = (item: ActionListData) => (
  <ActionListItemAction onClick={callFunction(item.columns[0]}}>
        Send Item: "{item detail here]}" to Someone... 
      </ActionListItemAction>
)
```

Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/useActionListSortManager.tsx]: Resolved in 011c45c!

File: packages/components/src/ActionList/utils/useActionListSortManager.tsx (hunk 9 above)
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/useActionListSortManager.tsx]: Resolved in 011c45c! (Still need to add in unit test file(s) though for stringComparitor and doDefaultSort)

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Resolved in 011c45c!

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Resolved in 5157b0973

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx [hunk 11]
```diff
@@ -23,15 +23,67 @@
...
+    const columnInfo = columns && columns.find(column => column.id === id)
+
+    const handleClick = () => {
+      if (doSort && columnInfo && columnInfo.canSort) {
+        if (columnInfo && columnInfo.sortDirection === 'desc') {
+          doSort(id, 'asc')
+        } else {
+          doSort(id, 'desc')
+        }
+      }
```
Comment (lukelooker) [inline packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]:
```suggestion
        doSort(id, columnInfo.sortDirection === 'asc' ? 'desc' : 'asc')
```

File: packages/components/src/ActionList/utils/doDefaultSort.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
+import { ActionListColumns } from '../ActionList'
+
+export const stringComparator = (stringA: string, stringB: string) => {
+  const upperCasedStringA = stringA.toUpperCase()
+  const upperCasedStringB = stringB.toUpperCase()
+
+  if (upperCasedStringA < upperCasedStringB) return -1
+  if (upperCasedStringA > upperCasedStringB) return 1
+  return 0
+}
```
Comment (lukelooker) [inline packages/components/src/ActionList/utils/doDefaultSort.ts]: I'm guessing this is exported so it can be tested? (which is fine) however if it's exported it will end up with the general name-space for imports so it needs to be more specifically named.

File: packages/components/src/ActionList/utils/doDefaultSort.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+export type ActionListDatum = Record<string, any>
+export type ActionListData = ActionListDatum[]
+
+export const doDefaultSort = (
```
Comment (lukelooker) [inline packages/components/src/ActionList/utils/doDefaultSort.ts]: Same thought as above... needs to be more explicitly named if this is getting exported.

One thought is that you could not do an `export * from './doDefaultSort'` from the index.ts and _just_ import / export the pieces that external consumers will need.

File: packages/components/src/ActionList/utils/doDefaultSort.ts (hunk 12 above)
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/doDefaultSort.ts]: Resolved in https://github.com/looker-open-source/components/pull/668/commits/a7fd571d27f12b50040aae4b034c5cdd9146b306

File: packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx (hunk 11 above)
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/ActionListHeader/ActionListHeaderColumn.tsx]: Resolved in https://github.com/looker-open-source/components/pull/668/commits/a7fd571d27f12b50040aae4b034c5cdd9146b306!

File: packages/components/src/ActionList/utils/doDefaultSort.ts (hunk 13 above)
Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/doDefaultSort.ts]: Resolved, _I think_, in https://github.com/looker-open-source/components/pull/668/commits/a7fd571d27f12b50040aae4b034c5cdd9146b306!

Regarding changes to the `ActionList/utils/index.ts` file, I created a more specific export statement that doesn't include the string comparison function.

Comment (ElliotTheHuman) [reply packages/components/src/ActionList/utils/doDefaultSort.ts]: Cleaned up a bit more in bca133e
